README.md
# usi-wordpress-solutions #

The WordPress-Solutions plugin is a helper class used by various WordPress plugins and themes developed by 
Universal Solutions.

## Installation ##
This module should be installed in the usi-wordpress-solutions folder under the main WordPress plugins folder. 
There are also some third party packages that must be installed if you use them:

### PHPSpreadsheet Installation
If you want to create and download Microsoft Excel documents then you have to install PHPOffice/PHPSpreadsheet 
which is a pure PHP library for reading and writing spreadsheet files. Go to the /PHPSpreadsheet 
folder and enter:
```
composer require phpoffice/phpspreadsheet
```
to get the latest version of PHPSpreadsheet. 
See [https://github.com/PHPOffice/PhpSpreadsheet](https://github.com/PHPOffice/PhpSpreadsheet) and [https://phpspreadsheet.readthedocs.io](https://phpspreadsheet.readthedocs.io) for more information.

PHPSpreadsheet is the next version of PHPExcel. 
It breaks compatibility to dramatically improve the code base quality 
(namespaces, PSR compliance, use of latest PHP language features, etc.).
Because all efforts have shifted to PhpSpreadsheet, PHPExcel will no longer be maintained. 

### PHPWord Installation
If you want to create and download Microsoft Word documents then you have to install PHPOffice/PHPWord 
which is a pure PHP library for reading and writing word processing documents. 
Go to the /extractions/phpword folder and enter:
```
composer require phpoffice/phpword
```
to get the latest version of PHPWordt. 
See [https://github.com/PHPOffice/PHPWord](https://github.com/PHPOffice/PHPWord) and [https://phpword.readthedocs.io](https://phpword.readthedocs.io) for more information.

### mPDF Installation
If you want to create and download PDF documents then you have to install mPDF. Go to the /extractions/mypdf folder and enter:
```
composer require mpdf/mpdf
```
to get the latest version of mpdf.

### TinyMCE Installation
Go to the /extractions/tinymce_5.7.1/tinymce/js/tinymce folder and copy everything to the tinyMCE folder.

## Debugging ##
Notes to help debug some WordPress issues.

### Deprecation errors with PHP version 8+
WordPress gives many deprecation errors when running in PHP version 8+ with debugging enabled. 
You can disable these errors but still enable debugging by doing the following:

```
in the wp-config.php file:

// Turn off default debugging:
define('WP_DEBUG', false); 

// Turn back on  debugging options without E_DEPRECATED;
error_reporting(E_ALL&~(E_DEPRECATED|E_USER_DEPRECATED));
$GLOBALS['wp_filter']=['enable_wp_debug_mode_checks'=>[10=>[['accepted_args'=>0,'function'=>function(){return(false);}]]]];
```

### Admin menu gap issue when debug mode is enabled
Sometimes when debug mode is enabled there is a 2em gap between WordPress admin menu and the bar.
This gap will show when there is a "Hidden" error if the following conditions are true:
```
if ($error && WP_DEBUG && WP_DEBUG_DISPLAY && ini_get('display_errors'))
```
To show the "Hidden" error do the following...
```
in the wp-config.php file:

define('WP_DEBUG', true);
define('WP_DEBUG_DISPLAY', false|true);
define('WP_DEBUG_LOG', true); // Or give the desired log file path; 

in the wp-admin/admin-header.php file around line 201:

$error = error_get_last();
error_log('=== This is the "Hidden" error =================');
error_log(print_r($error, true));
error_log('=== Error Setting display ======================');
error_log('WP_DEBUG');
error_log(print_r(WP_DEBUG, true));
error_log('WP_DEBUG_DISPLAY');
error_log(print_r(WP_DEBUG_DISPLAY, true));
error_log('display_errors ini setting');
error_log(print_r(ini_get('display_errors'), true));
```
The default location of the debug.log file is in the wp-content folder.


## Pop Ups ##
This plugin provides enhanced pop functionality.
```
class USI_WordPress_Solutions_Popup_Action is used in:
 * ru-application.php
 * ru-docusign.php
 * ru-form.php
 * ru-work.php
 * usi-media-solutions-manage.php

class USI_WordPress_Solutions_Popup_Iframe is used in:
 * ru-email-inbox.php
 * usi-test-solutions-settings.php
 * usi-wordpress-solutions-versions.php
 * usi-wordpress-solutions-settings-settings.php
```


## License ##
> WordPress-Solutions is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License 
as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

> WordPress-Solutions is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

> You should have received a copy of the GNU General Public License along with Variable-Solutions.  If not, see 
<http://www.gnu.org/licenses/>.

## Donations ##
Donations are accepted at <a href="https://www.usi2solve.com/donate/wordpress-solutions">www.usi2solve.com/donate</a>. Thank you for your support!

CHANGELOG.md
# Change Log #

WordPress-Solutions plugin changes are logged here using <a href="http://semver.org/">Semantic Versioning</a>.

## 2.16.3 (2023-11-27) ##
* Modified to use proper coding constructs.

## 2.16.2 (2023-10-10) ##
* Added custom code execution via short code.

## 2.16.1 (2023-09-26) ##
* Added temp directory for mPDF.

## 2.16.0 (2023-09-15) ##
* Added autoloader for USI plugins.
* Set all versions numbers to same version.

## 2.15.5 (2023-07-07) ##
* Added styling to trasnfer page.

## 2.15.4 (2023-07-06) ##
* Added common WordPress helper functions for returning filter values.

## 2.15.3 (2023-07-03) ##
* Moved the usi::log() method out of the plugin and into it's own must-use plugin.

## 2.15.2 (2023-07-02) ##
* Added disable visual editing in classic editor option.

## 2.15.1 (2023-06-30) ##
* Removed admin functions from main file and put into admin file.

## 2.15.0 (2023-06-30) ##
* Set all versions numbers to same version.

## 2.14.11 (2023-06-20) ##
* Fixed popup handling, added e-mail cloaking.

## 2.14.10 (2023-05-22) ##
* Improved popup handling.

## 2.14.9 (2023-05-19) ##
* Corrected minor typos and issue with plugin version display.

## 2.14.8 (2023-03-16) ##
* Modified so that missings settings sections doesn't throw error.

## 2.14.7 (2023-03-15) ##
* Added PDF watermark, updated code comments.

## 2.14.6 (2023-02-07) ##
* Upgraded the mPDF system to version 8.1.4.

## 2.14.5 (2022-12-19) ##
* Modifications made for PHP 8.1 compatibility.

## 2.14.4 (2022-11-01) ##
* Modifications made for PHP 8.1 compatibility.

## 2.14.3 (2022-10-05) ##
* Consolidated jQuery script handling.

## 2.14.2 (2022-09-23) ##
* Improved xPorter functionality to Transfer.

## 2.14.1 (2022-08-10) ##
* Adding PHPMailer support.
* Set all versions numbers to same version.

## 2.14.0 (2022-06-19) ##
* Set all versions numbers to same version.

## 2.13.1 (2022-03-17) ##
* Added more illumination variables.

## 2.13.0 (2022-02-22) ##
* Added first pass for a form designer/layout tool.
* Set all versions numbers to same version.

## 2.12.15 (2022-02-03) ##
* Added more illumination variables.

## 2.12.14 (2022-02-01) ##
* Added more illumination variables.

## 2.12.13 (2022-01-30) ##
* Added more illumination variables.

## 2.12.12 (2022-01-26) ##
* Added more illumination variables.

## 2.12.11 (2022-01-25) ##
* Added illumination variables.

## 2.12.10 (2022-01-24) ##
* Enhanced security.

## 2.12.9 (2022-01-23) ##
* Modified so that bad folder does not give errors.

## 2.12.8 (2022-01-21) ##
* Modified so that setting changes the pcre.backtrack_limit value for PDF downloads.

## 2.12.7 (2022-01-19) ##
* Added setting to set the pcre.backtrack_limit value.

## 2.12.6 (2022-01-18) ##
* Updated to print error message in the PDF file when PDF creation fails.

## 2.12.5 (2022-01-16) ##
* phpinfo() scan now requires administrator privileges.

## 2.12.4 (2022-01-15) ##
* Added USI_WordPress_Solutions_Logged_In class which uses cookies to verify if user is logged in.

## 2.12.3 (2022-01-14) ##
* Added bootstrap to load WordPress functions from an external .PHP file.

## 2.12.2 (2021-11-29) ##
* Apdated settings functions.

## 2.12.1 (2021-11-16) ##
* Used ?? operator in get_value() function.

## 2.12.0 (2021-11-03) ##
* Set all versions numbers to same version.

## 2.11.17 (2021-10-01) ##
* Added USI_WordPress_Solutions_PDF class.
* Fixed some PHP warning messages.

## 2.11.16 (2021-07-23) ##
* Added $active_tab_maybe variable.

## 2.11.15 (2021-07-14) ##
* Added headers and footers to PDF dumps.

## 2.11.14 (2021-06-24) ##
* Addressed an issue where checked check boxes couldn't be cleared.

## 2.11.13 (2021-06-16) ##
* Added check in processing settings.

## 2.11.12 (2021-06-15) ##
* Added clear function to datepicker widget.

## 2.11.11 (2021-06-11) ##
* Changed usi::log() to accomondate the Page Solutions plugin.

## 2.11.10 (2021-06-10) ##
* Set jQuery datepicker plugin date format.

## 2.11.9 (2021-06-09) ##
* Fixed some PHP warnings resulting from new PHP version.
* Re-ordered some functions alphavetically.

## 2.11.8 (2021-05-29) ##
* Added datepicker support.
* Modified to ensure backwards compatibility with PHP 5.4 versions.

## 2.11.7 (2021-05-25) ##
* Added PHP error reporting function to settings, addressed false errors in mPDF creation.

## 2.11.6 (2021-05-18) ##
* Added feature to remove the password reset link option from the row actions in the user display.

## 2.11.5 (2021-05-12) ##
* Added .usi-ignore file to skip sub folder from version scanning in the all-scan module.

## 2.11.4 (2021-05-06) ##
* Added e-mail field type and validation function.
* Added .usi-ignore file to skip sub folder from version scanning.

## 2.11.3 (2021-04-20) ##
* Set all versions numbers to same version.
* Added TinyMCE editor to settings functionality.
* Increseaed the size of the USI::log() action field from TEXT to MEDIUMTEXT.
* Made USI::log() class protected by if defined logic so that file can be included anywhere.

## 2.11.2 (2021-03-21) ##
* Moved section_header() so that all sub-classes can utilize.

## 2.11.1 (2021-03-10) ##
* Changed esc_attr() to esc_textarea() for text area settings fields.

## 2.11.0 (2021-02-24) ##
* Set all versions numbers to same version.

## 2.10.7 (2021-02-02) ##
* Added ability to force user logout from active sessions list.

## 2.10.6 (2021-01-20) ##
* Added diagnostic logging to post export feature.

## 2.10.5 (2021-01-04) ##
* Converted the USI_WordPress_Solutions_Settings::action_init() function to be static so it could be called by other plugins to impersonate a user.

## 2.10.4 (2020-12-21) ##
* Improved money field functionality.

## 2.10.3 (2020-12-04) ##
* Added remove_directory() method to the USI_WordPress_Solutions_Static class.

## 2.10.2 (2020-11-18) ##
* Added eXporter to settings to export posts, updated popup usage list, added money settings field type.

## 2.10.1 (2020-11-02) ##
* Added USI_WordPress_Solutions_Popup_Action class.

usi-wordpress-solutions-popup.php
<?php // ------------------------------------------------------------------------------------------------------------------------ //

defined('ABSPATH') or die('Accesss not allowed.');

/*
WordPress-Solutions is free software: you can redistribute it and/or modify it under the terms of the GNU General Public 
License as published by the Free Software Foundation, either version 3 of the License, or any later version.
 
WordPress-Solutions is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied 
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 
You should have received a copy of the GNU General Public License along with WordPress-Solutions. If not, see 
https://github.com/jaschwanda/wordpress-solutions/blob/master/LICENSE.md

Copyright (c) 2020 by Jim Schwanda.
*/

// https://codex.wordpress.org/Javascript_Reference/ThickBox
// http://codylindley.com/thickbox/

/*
accept  = string|null                   // Accept button text, left most button user clicks if they accept/confirm the popup operation;
action  = string|null                   // WP_List_Table:item mouse over link action;
body    = string|null                   // Text message body for inline popups, automatically generated for WP_List_Table;
bulk    = string|'[]'                   // WP_List_Table:string representation of posible bulk actions array;
cancel  = string|null                   // Cancel button text, second button from left user clicks if they cancel the popup operation;
choice  = string|null                   // WP_List_Table:message displayed if no selection made prior to invoking popup;
class   = string                        // html class added to anchor, usualy used to identofy items to be scanned for selection;
close   = string|null                   // Close button text, usually used to remove information only popups;
direct  = string|null                   // jQuery selector(s) that when clicked start scan of items in document to include in popup operation, must include . and # and may include more than one;
format  = 'br'|'li'                     // List format, use breaks or list tags;
id      = string|null                   // Id of hidden <div> that contains the popup elements;
info    = string|null                   // Information included in the data-info"" attribute for item anchor;
input   = 'file'|'check'                // Input field types, usually checkboxes;
key     = string|null                   // Key included in the data-key"" attribute that uniquely identifies an item anchor;
page    = string|null                   // WP_List_Table:page that contains the list table, used to execute the action if popup is confirmed;
list    = string|null                   // jQuery selector(s) that select the items to include in the popup operation, must include . and # and may include more than one;
link    = string|null                   // Displayed text for an item anchor, what goes between the <a></a> tags;
modal   = true|null                     // Causes popup window to be modal, the title bar and close X are not show, clicking outside the popup does not close the popup;
ok      = string|null                   // OK button text, usually used to remove instruction only popups, like "please make a swlwction...", and user clicks the OK button;
pass    = int|null                      // Number of times or "passes" the build popup builder is called, null means only one time, positive numbers indicate the pass number, inline and script html are only generated on the 1st pass;
prefix  = string|null                   // Text string displayed before body text;
select  = string|null                   // WP_List_Table:message displayed if no bulk action selected prior to invoking popup;
submit  = string|null                   // jQuery selector string that invokes popup and submits the form after popup confirmation;
suffix  = string|null                   // Text string displayed after body text;
table   = true|false                    // The popup is to confirm an individual or bulk WP_List_Table action;
tip     = string|null                   // Tool tip that shows when you hover over an anchor link;
title   = string|'WordPress-Solutions'  // popup title;
type    = 'ajax'|'iframe'|'inline'      // popup implementation type;
url     = sting|null                    // URL for iframe source;
width   = int|300                       // popup width in pixels;


Popup with inline window;
Popup with iframe and load another file;
Popup with where list a itmes listed asking for confirmation;

*/

class USI_WordPress_Solutions_Popup {

   const VERSION = '2.14.3 (2022-10-05)';

   private function __construct() {
   } // __construct();

   public static function build($options) {

      if (!empty($options['iframe'])) {
         die(__METHOD__.':'.__LINE__.':not supported');
      }

      $accept = !empty($options['accept']) ?          $options['accept']  : null;
      $action = !empty($options['action']) ?          $options['action']  : null;
      $body   = !empty($options['body']  ) ?          $options['body']    : null;
      $bulk   = !empty($options['bulk']  ) ?          $options['bulk']    : '[]';
      $cancel = !empty($options['cancel']) ?          $options['cancel']  : null;
      $choice = !empty($options['choice']) ?          $options['choice']  : null;
      $class  = !empty($options['class'] ) ? esc_attr($options['class'])  : null;
      $close  = !empty($options['close'] ) ?          $options['close']   : null;
      $direct = !empty($options['direct']) ? esc_attr($options['direct']) : null;
      $format = !empty($options['format']) ?          $options['format']  : 'br';
      $height = !empty($options['height']) ?     (int)$options['height']  : 300;
      $id     = !empty($options['id']    ) ?          $options['id']      : null;
      $key_id = !empty($options['key_id']) ?          $options['key_id']  : null;
      $info   = !empty($options['info']  ) ? esc_attr($options['info'])   : null;
      $key    = !empty($options['key']   ) ? esc_attr($options['key'])    : null;
      $link   = !empty($options['link']  ) ? esc_attr($options['link'])   : null;
      $list   = !empty($options['list']  ) ?          $options['list']    : null;
      $modal  = !empty($options['modal'] ) ?         ($options['modal'])  : null;
      $ok     = !empty($options['ok']    ) ?          $options['ok']      : null;
      $page   = !empty($options['page']  ) ?          $options['page']    : null;
      $pass   = !empty($options['pass']  ) ?     (int)$options['pass']    : 0;
      $prefix = !empty($options['prefix']) ?          $options['prefix']  : null;
      $select = !empty($options['select']) ?          $options['select']  : null;
      $submit = !empty($options['submit']) ?          $options['submit']  : null;
      $suffix = !empty($options['suffix']) ?          $options['suffix']  : null;
      $table  = !empty($options['table'] ) ?          $options['table']   : null;
      $tip    = !empty($options['tip']   ) ? esc_attr($options['tip'])    : null;
      $title  = !empty($options['title'] ) ? esc_attr($options['title'])  : 'WordPress-Solutions Popup';
      $type   = !empty($options['type']  ) ?          $options['type']    : 'inline';
      $url    = !empty($options['url']   ) ?          $options['url']     : null;
      $width  = !empty($options['width'] ) ?     (int)$options['width']   : 300;

      $anchor = $inline = $script = $submit1 = null;
      $offset = 72;

      $iframe = 'iframe' == $type;

      if ($link) {
         $anchor = '<a class="' . $class . '"' .
            ($action && $pass ? ' data-action="'  . $action  . '"' : '') .
            ($info   && $pass ? ' data-info="'    . $info  . '"' : '') .
            ($key    && $pass ? ' data-key="'     . $key   . '"' : '') .
            ($iframe && $pass ? ' data-title="'   . $title . '"' : '') .
            ($iframe && $pass ? ' data-url="'     . $url   . '"' : '') . 
            ' href=""' . ($tip ? ' title="' . $tip . '"' : '') . '>' . $link . '</a>';
      }

      if (true === $modal) {
         $modal   = '&modal=true';
         $offset -= 29;
      }

      if ($submit) $submit1 = explode(',', $submit)[0];

      $table = $table ? 'true' : 'false';

      if (1 >= $pass) {

         if ($id) $inline = '<div id="' . $id . '" style="display:none"></div><!--' . $id . '-->';

         if ($pass) $url = null;

         $script = <<<EOD
// BEGIN - {$id}
var is_active = false;
var is_iframe = false;

var log    = true;

var accept = '{$accept}';
var body   = '{$body}';
var cancel = '{$cancel}';
var choice = '{$choice}';
var close  = '{$close}';
var format = '{$format}';
var ok     = '{$ok}';
var prefix = '{$prefix}';
var select = '{$select}';
var suffix = '{$suffix}';
var title  = '{$title}';
var type   = '{$type}';
var url    = '{$url}';

var show_accept = false;
var show_body   = false;
var show_cancel = false;
var show_close  = false;
var show_ok     = false;
var show_prefix = false;
var show_select = false;
var show_suffix = false;

var action      = null;
var html        = '';
var id_list     = null;
var style       = '';
var table       = {$table};

// Get bulk action for table confirmation popups;
function bulk() {
   var actions = {$bulk};
   var top     = $('#bulk-action-selector-top').val();
   var bot     = $('#bulk-action-selector-bottom').val();
   if (top || bot) {
      for (var i  = 0; i < actions.length; i++) {
         if (top == actions[i]) return(top);
         if (bot == actions[i]) return(bot);
      }
   }
   // RETURN with direct action if any;
   return(action);
} // bulk();

function hide(text) {
   // Use <span> since not invoking link to another page;
   html += '<span class="button" onclick="tb_remove();"' + style + '>' + text + '</span> &nbsp;';
   style = '';
} // hide():

function init() {
   action      = null;
   show_accept = 0 < accept.length;
   show_body   = 0 < body.length;
   show_cancel = 0 < cancel.length;
   show_close  = 0 < close.length;
   show_ok     = 0 < ok.length;
   show_prefix = 0 < prefix.length;
   show_select = 0 < select.length;
   show_suffix = 0 < suffix.length;
   // Style pads actions buttons at bottom of iframe to look like inline popup;
   style       = ' style="margin-left:15px;"'
} // init();

function show() {

   if (log) console.log('show:begin');

   // Addresses bug in thickbox that doesn't allow height/wdith option on some admin pages;
   var old_position = tb_position;
   tb_position = function() {
      $('#TB_window').css(
         {
            height : TB_HEIGHT + 'px', 
            marginLeft : '-' + parseInt((TB_WIDTH / 2), 10) + 'px', 
            marginTop : '-' + parseInt((TB_HEIGHT / 2), 10) + 'px',
            width : TB_WIDTH + 'px', 
         }
      );
   }; // tb_position();

   if ('iframe' == type) {
      is_iframe = show_cancel || show_close || show_ok;
      tb_show(title, url + '?TB_iframe=true&height={$height}{$modal}&width={$width}', null);
   } else if ('inline' == type) {
      tb_show(title, 'TB_inline?inlineId={$id}&height={$height}{$modal}&width={$width}', null);
      style = '';
      $('#TB_ajaxContent').html(view());
      work();
   }

   tb_position = old_position;

   if (log) console.log('show:end');

} // show();

function view() {

   if (log) console.log('view:begin');

   html  = '';
   if (show_prefix) html += '<p>' + prefix + '</p>';
   if (show_body)   html += (('li' == format) ? '<ul>' : '') + body   + (('li' == format) ? '</ul>' : '');
   if (show_suffix) html += '<p>' + suffix + '</p>';
   html += '<hr><p>';
   if (show_accept) {
      if (table) {
         action = bulk();
         html  += '<a class="button" href="?page={$page}&action=' + action + '&{$key_id}=' + id_list + '">' + accept + '</a> &nbsp; ';
      } else {
         html  += '<span id="usi-wordpress-solutions-popup-confirm" class="button">' + accept + '</span> &nbsp; ';
      }
      style     = '';
   }
   if (show_cancel) hide(cancel);
   if (show_ok)     hide(ok);
   if (show_close)  hide(close);
   html += '</p>';

   if (log) console.log('view:end:html=' + html);

   return(html);

} // view();

// Re-submit form, this function must be called after accept button loaded into DOM;
function work() {
   $('#usi-wordpress-solutions-popup-confirm').click(
      () => {
         is_active = true;
         tb_remove();
         $('{$submit1}').click();
         return(false);
      }
   );
} // work();

// Extend iframe thickbox to hold buttons at bottom;
$('body').on(
   'thickbox:iframe:loaded', 
   () => {
      if (is_iframe) {
         is_iframe = false;
         $('#TB_iframeContent').height($('#TB_iframeContent').height() - {$offset});
         $('#TB_iframeContent').after(view());
         work();
      }
   }
);

// Invoke popup indirectly by clicking on submit button that scans document for elements;
$('{$submit}').click(
   (event) => {

      if (log) console.log("$('{$submit}').click():begin");

      if (is_active) {
         is_active = false;
         if (log) console.log("$('{$submit}').click():is_active:end");
         return(true);
      }

      init();

      var t = event.target;
      if (t.getAttribute('data-title')) title = t.getAttribute('data-title');
      if (t.getAttribute('data-url')  ) url   = t.getAttribute('data-url');

      if ('inline' == type) {
         action    = bulk();
         if (table && !action) {
            body   = select;
            show_prefix = show_suffix = show_cancel = show_close = show_accept = false;
         } else {
            body    = '';
            id_list = '';
            var ids = $('{$list}');
            if (log) console.log("$('{$submit}').click():$('{$list}').length=" + ids.length);
            try {
            for (var i = 0; i < ids.length; i++) {
               if (ids[i].checked) {
                  var key  = ids[i].getAttribute('data-key');
                  var info = ids[i].getAttribute('data-info');
                  if (!key && !info) {
                     //Take attributes from anchor tag in next field;
                     key  = ids[i].parentElement.nextSibling.firstChild.getAttribute('data-key');
                     info = ids[i].parentElement.nextSibling.firstChild.getAttribute('data-info');
                  }
                  if (log) console.log("$('{$submit}').click():$('{$list}').key=" + key + ' info=' + info);
                  id_list += (id_list.length ? ',' : '')  + key;
                  if ('li' == format) {
                     body    += '<li>' + info + '</li>';
                  } else {
                     body    += (body ? '<br/>' : '') + info;
                  }
               } else if (ids[i].files && ids[i].files[0] && ids[i].files[0].name) {
                  var name = ' &nbsp; ' + ids[i].files[0].name;
                  if (log) console.log("$('{$submit}').click():$('{$list}').name=" + name);
                  if ('li' == format) {
                     body    += '<li>' + name + '</li>';
                  } else {
                     body    += (body ? '<br/>' : '') + name;
                  }
               }
            }
            } catch (exception) {
               console.log(exception.message);
            }
            if (body) {
               show_ok = false;
            } else {
               body    = choice;
               show_prefix = show_suffix = show_cancel = show_close = show_accept = false;
            }
         }
         show_body = 0 < body.length;
      }

      show();

      if (log) console.log("$('{$submit}').click():end");

      return(false);

   }
);

// Invoke popup directly by click on element;
$('{$direct}').click(
   function(event) {

      if (log) console.log("$('{$direct}').click():begin");

      init();

      var t = event.target;
      if (t.getAttribute('data-action')) action = t.getAttribute('data-action');
      if (t.getAttribute('data-title'))  title  = t.getAttribute('data-title');
      if (t.getAttribute('data-url')  )  url    = t.getAttribute('data-url');

      if ('inline' == type) {
         body      = t.getAttribute('data-info');
         show_body = 0 < body.length;
         show_ok   = false;
         id_list   = t.getAttribute('data-key');
      }

      show();

      if (log) console.log("$('{$direct}').click():end");

      return(false);

   }
); // $('{$direct}').click();
// END - {$id}

EOD;

         USI_WordPress_Solutions::admin_footer_jquery($script);

      }

      return(array('anchor' => $anchor, 'inline' => $inline));

   } // build();

} // Class USI_WordPress_Solutions_Popup;

// --------------------------------------------------------------------------------------------------------------------------- // ?>